Add support for Python 3 (3.3, 3.4, 3.5)

MatthieuDartiailh opened the pull request:
This makes atom compatible with Python 3 in a python 2 compatible way : 
- it adds future as a dependency (needed for metaclass support, int types, ...)
- I added __future__ imports to all files which may not always be relevant I can remove those if you want. 
- @peterazmanov left some comments in atomlist, I can address them given some pointers.
- Python version checks in C code uses PY_MAJOR_VERSION and PY_MINOR_VERSION rather than the hex representation. @jwiggins used the hex in kiwi. I can move to the hex representation if necessary.
- all iteritems have been removed in favor of items, if performance is a concern for Python 2, future provides a cross-compatible shim that could be used instead.

sccolbert commented:
This will take some time for me to go through and comment on. I will leave
a note indicating when I'm done commenting. Thanks!

On Sun, Mar 13, 2016 at 11:44 AM, Matthieu Dartiailh <
notifications@github.com> wrote:

> This makes atom compatible with Python 3 in a python 2 compatible way :
>
>    - it adds future as a dependency (needed for metaclass support, int
>    types, ...)
>    - I added *future* imports to all files which may not always be
>    relevant I can remove those if you want.
>    - @peterazmanov <https://github.com/peterazmanov> left some comments
>    in atomlist, I can address them given some pointers.
>    - Python version checks in C code uses PY_MAJOR_VERSION and
>    PY_MINOR_VERSION rather than the hex representation. @jwiggins
>    <https://github.com/jwiggins> used the hex in kiwi. I can move to the
>    hex representation if necessary.
>    - all iteritems have been removed in favor of items, if performance is
>    a concern for Python 2, future provides a cross-compatible shim that could
>    be used instead.
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/nucleic/atom/pull/65
> Commit Summary
>
>    - Initial Python3 Support.
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 Support
>    - Initial Python3 support
>    - 2to3
>    - introduce Bytes atom-type, make Str alias of Unicode
>    - sortedmap quick fix (enaml 0.9.8 uses unorderable in py3k keys)
>    - fix merge issues
>    - Update all python modules for python 2/3 compatibilty.
>    - Add future as a dependence.
>    - Update src for Python 2/3 compatibility.
>    - Update examples and tests.
>    - Fix compilation.
>    - Fix bug in scalars and bump version
>    - Add missing modes to validate modes.
>    - Drop use of unicode_litterals inside atom as it caused more issues
>    than anything else.
>    - Add support for python 3.5 (based on
>    https://github.com/nucleic/atom/pull/64)
>    - Merge pull request #1 from MatthieuDartiailh/py23
>    - fix: change tabs to spaces
>    - fix bug: no IS_PY3K is defined
>    - setup.py: add code to simplify extension build in debug mode
>    - py3k: remove cmp keyword from atomlist::sort arguments, update tests
>    - add '/.idea' to gitignore
>    - setup : Fix use of super with distutils build_ext
>    - release some type constraints on Python 2
>    - minor clean up
>
> File Changes
>
>    - *M* .gitignore <https://github.com/nucleic/atom/pull/65/files#diff-0>
>    (1)
>    - *M* atom/api.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-1> (7)
>    - *M* atom/atom.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-2> (36)
>    - *M* atom/coerced.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-3> (2)
>    - *M* atom/containerlist.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-4> (2)
>    - *M* atom/delegator.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-5> (2)
>    - *M* atom/dict.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-6> (15)
>    - *M* atom/enum.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-7> (2)
>    - *M* atom/event.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-8> (2)
>    - *M* atom/instance.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-9> (2)
>    - *M* atom/intenum.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-10> (19)
>    - *M* atom/list.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-11> (2)
>    - *M* atom/property.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-12> (2)
>    - *M* atom/scalars.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-13> (34)
>    - *M* atom/signal.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-14> (2)
>    - *M* atom/src/atomlist.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-15> (118)
>    - *M* atom/src/atomref.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-16> (35)
>    - *M* atom/src/behaviors.h
>    <https://github.com/nucleic/atom/pull/65/files#diff-17> (6)
>    - *M* atom/src/catom.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-18> (26)
>    - *M* atom/src/catommodule.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-19> (49)
>    - *M* atom/src/defaultvaluebehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-20> (3)
>    - *M* atom/src/delattrbehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-21> (7)
>    - *M* atom/src/enumtypes.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-22> (13)
>    - *M* atom/src/enumtypes.h
>    <https://github.com/nucleic/atom/pull/65/files#diff-23> (4)
>    - *M* atom/src/eventbinder.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-24> (16)
>    - *M* atom/src/getattrbehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-25> (9)
>    - *M* atom/src/member.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-26> (46)
>    - *M* atom/src/memberchange.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-27> (21)
>    - *M* atom/src/methodwrapper.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-28> (42)
>    - *M* atom/src/postgetattrbehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-29> (3)
>    - *M* atom/src/postsetattrbehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-30> (3)
>    - *M* atom/src/postvalidatebehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-31> (3)
>    - *A* atom/src/py23compat.h
>    <https://github.com/nucleic/atom/pull/65/files#diff-32> (54)
>    - *M* atom/src/pythonhelpers.h
>    <https://github.com/nucleic/atom/pull/65/files#diff-33> (32)
>    - *M* atom/src/setattrbehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-34> (11)
>    - *M* atom/src/signalconnector.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-35> (15)
>    - *M* atom/src/sortedmap.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-36> (91)
>    - *M* atom/src/utils.h
>    <https://github.com/nucleic/atom/pull/65/files#diff-37> (14)
>    - *M* atom/src/validatebehavior.cpp
>    <https://github.com/nucleic/atom/pull/65/files#diff-38> (93)
>    - *M* atom/subclass.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-39> (2)
>    - *M* atom/tuple.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-40> (2)
>    - *M* atom/typed.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-41> (2)
>    - *M* atom/version.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-42> (4)
>    - *M* examples/api/coersion.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-43> (3)
>    - *M* examples/api/composition.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-44> (3)
>    - *M* examples/api/containers.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-45> (3)
>    - *M* examples/api/default_value.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-46> (4)
>    - *M* examples/api/numeric.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-47> (6)
>    - *M* examples/api/observe.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-48> (3)
>    - *M* examples/api/property.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-49> (3)
>    - *M* examples/api/range.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-50> (3)
>    - *M* examples/tutorial/employee.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-51> (4)
>    - *M* examples/tutorial/hello_world.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-52> (3)
>    - *M* examples/tutorial/person.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-53> (3)
>    - *M* setup.py <https://github.com/nucleic/atom/pull/65/files#diff-54>
>    (10)
>    - *M* tests/test_atomlist.py
>    <https://github.com/nucleic/atom/pull/65/files#diff-55> (181)
>
> Patch Links:
>
>    - https://github.com/nucleic/atom/pull/65.patch
>    - https://github.com/nucleic/atom/pull/65.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nucleic/atom/pull/65>.
>

MatthieuDartiailh commented:
No problem. I will try to make the enaml PRs more manageable. Note that the kiwi PR (https://github.com/nucleic/kiwi/pull/22) is ready too.

MatthieuDartiailh commented:
Any progress in the review @sccolbert ?

jwiggins commented:
IIRC, I used the hex version to avoid this particular situation:
```
#if PY_MAJOR_VERSION >= 3
#if PY_MINOR_VERSION > 4
    ( PyAsyncMethods* )0,                  /* tp_as_async */
#else
    ( void* ) 0,                           /* tp_reserved */
#endif
#else
...
#endif
```
I believe my initial reasoning what that it looked a bit crowded, but looking at it again I see another problem. If Python's major version is ever incremented, the minor version will no longer be greater than 4, and then you'll switch back to this being a reserved slot. Probably OK, but certainly not your intent.

MatthieuDartiailh commented:
Makes sense @jwiggins. I found the hex not super readable but your argument convinced me it is indeed the right thing to do. I will update the code accordingly.

MatthieuDartiailh commented:
The object declaration now use the PY_VERSION_HEX as suggested by @jwiggins

MatthieuDartiailh commented:
ping @sccolbert

sccolbert commented:
Thanks. This is all very appreciated!!! I'm just slammed with other work atm. Sorry :(

blink1073 commented:
Chris is just lazy, don't listen to him /s.

MatthieuDartiailh commented:
@sccolbert any hope that you will be able to give it a first review soon ?

MatthieuDartiailh commented:
@sccolbert any hope you could give this a quick look ?

MatthieuDartiailh commented:
Sorry for bothering you again, but do you have time to give this a look ?

MatthieuDartiailh commented:
The last commit fix the long standing bug seen in enaml on Python 3 https://github.com/nucleic/enaml/issues/216

jwiggins commented on atom/api.py:
```diff
@@ -1,10 +1,13 @@
 # The full license is in the file COPYING.txt, distributed with this software.
 #------------------------------------------------------------------------------
+from __future__ import (division, unicode_literals, print_function,
+                        absolute_import)
```
I noticed that you have this in every module in the `atom` package. However I don't think it's needed:
* nothing is using the divide operator
* nothing is printing (examples are another story...)
* `absolute_import`: optional after Python 2.5.0a1, and this project already uses absolute imports
* `unicode_literals`: optional after Python 2.6.0a2, only used in `scalars.py`

See: https://docs.python.org/2/library/__future__.html

I'm not sure what versions of Python should be supported by atom, but it might be OK to not worry about 2.5 & 2.6 support.

MatthieuDartiailh replied:
Usually I prefer to systematically include those imports to be sure nothing breaks unexpectedly. But it add noise to this PR so I removed them as there are not exactly necessary.

jwiggins commented on atom/scalars.py:
```diff
@@ -176,21 +178,35 @@ def __init__(self, default=0.0, factory=None, strict=False):
-    By default, unicode strings will be promoted to plain strings. Pass
-    strict=True to the constructor to enable strict string checking.
+class Bytes(Value):
+    """ A value of type `bytes`.
```
You should really comment on the promotion rules here since they also affect this type.

jwiggins commented on atom/scalars.py:
```diff
@@ -202,7 +218,7 @@ class Unicode(Value):
     __slots__ = ()
 
-    def __init__(self, default=u'', factory=None, strict=False):
+    def __init__(self, default='', factory=None, strict=False):
```
I would leave the `u''`

jwiggins commented on atom/atom.py:
```diff
@@ -38,7 +47,7 @@ def observe(*names):
     pairs = []
     for name in names:
-        if type(name) is not str:
+        if not isinstance(name, basestring):
```
This also lets bytestrings through. Is that your intent?

jwiggins commented on atom/atom.py:
```diff
@@ -194,7 +203,7 @@ def __new__(meta, name, bases, dct):
         seen_sentinels = set()      # The set of seen sentinels
         seen_decorated = set()      # The set of seen @observe decorators
-        for key, value in dct.iteritems():
+        for key, value in dct.items():
```
I know you mentioned this in the intro for this PR, but I thought I would ask again: If the original version was using `iteritems`, don't you think it's a good idea to preserve that behavior?

I intend for this comment to cover all instances of `iteritems`->`items` and `itervalues`->`values`, but I'll spare you the repeated callouts.

MatthieuDartiailh replied:
This is debatable. I prefer to have a Python 3 oriented code base. Actually a similar choice was done in Chaco recently.

jwiggins commented on atom/coerced.py:
```diff
@@ -17,7 +19,8 @@ class Coerced(Member):
 
-    def __init__(self, kind, args=None, kwargs=None, factory=None, coercer=None):
+    def __init__(self, kind, args=None, kwargs=None, factory=None,
+                 coercer=None):
```
Technically, nothing _needed_ to be changed in this module

jwiggins commented on atom/intenum.py:
```diff
@@ -1,13 +1,23 @@
+else:
+    import copy_reg as copyreg
+
+from future.builtins import int as il
```
This name is a bit too terse

jwiggins commented on atom/src/pythonhelpers.h:
```diff
@@ -719,7 +743,7 @@ class PyMethodPtr : public PyObjectPtr {
     PyObjectPtr get_class() const
     {
-        return PyObjectPtr( PythonHelpers::xnewref( PyMethod_GET_CLASS( m_pyobj ) ) );
+        return PyObjectPtr( PythonHelpers::xnewref( (PyObject *)Py_TYPE( PyMethod_GET_SELF( m_pyobj ) ) ) );
```
Shouldn't this remain `PyMethod_GET_CLASS( m_pyobj )`?

jwiggins commented on atom/src/sortedmap.cpp:
```diff
@@ -278,17 +277,26 @@ SortedMap_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
+        // Clearing the vector may cause arbitrary side effects on item
+        // decref, including calls into methods which mutate the vector.
+        // To avoid segfaults, first make the vector empty, then let the
+        // destructors run for the old items.
```
Maybe move this comment up above the `#if` since it applies to both implementations?

jwiggins commented on atom/src/validatebehavior.cpp:
```diff
@@ -244,10 +244,12 @@ long_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject* newvalu
+    #if PY_MAJOR_VERSION < 3
+        if( PyInt_Check( newvalue ) )
+            return PyInt_FromLong( PyInt_AS_LONG( newvalue ) );
+    #endif
```
You're leaving this as Python 2 only, but you appear to be changing the behavior.

jwiggins commented on atom/src/validatebehavior.cpp:
```diff
@@ -266,8 +268,10 @@ float_promote_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject
+    #if PY_MAJOR_VERSION < 3
     if( PyInt_Check( newvalue ) )
         return PyFloat_FromDouble( static_cast<double>( PyInt_AS_LONG( newvalue ) ) );
+    #endif
```
Why is this being removed from Python 3?

MatthieuDartiailh replied:
PyInt_Check does not exist under Python 3 only PyLong_Check

jwiggins commented on atom/src/validatebehavior.cpp:
```diff
@@ -278,33 +282,58 @@ float_promote_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject
     if( PyUnicode_Check( newvalue ) )
         return newref( newvalue );
-    return validate_type_fail( member, atom, newvalue, "unicode" );
+    return validate_type_fail( member, atom, newvalue, "str" );
```
Why the change from `"unicode"` to `"str"`? (also below)

MatthieuDartiailh replied:
Addressed in a version specific manner now. (untetested)

jwiggins commented on atom/src/validatebehavior.cpp:
```diff
@@ -278,33 +282,58 @@ float_promote_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject
     if( PyUnicode_Check( newvalue ) )
         return PyUnicode_AsUTF8String( newvalue );
+
+#endif
```
The promotion rules for string types should be clarified.

MatthieuDartiailh replied:
Is not the Python side clear enough ?

jwiggins replied:
Not quite, no. The docstring for `Str` says:

>Under Python 2 this is a byte string, under Python 3 a unicode one.
>
>The use of this member is discouraged in Python 2/3 compatible codebase
>as Bytes and Unicode provide a more homogeneous behavior.

The other types describe what their promotion semantics are.

The behavior I see here:

INPUT | Unicode | Bytes
----- | ------- | -----
Py2 |  UTF-8 bytestring | Bytes
Py3 | Unicode | Unicode

jwiggins commented on atom/version.py:
```diff
@@ -22,7 +24,7 @@
 # The micro release number. The micro release number is incremented
 # for bug fix releases and small feature additions.
-MICRO = 10
+MICRO = 11
```
You think maybe a minor version bump might be in order?

MatthieuDartiailh replied:
This will be a minor version bump.

MatthieuDartiailh commented:
Thanks for the review @jwiggins. I am in the peogress of cleaning this up and your comments are very valuable. I plan to add more tests and enable Travis either in this PR or right after.

sccolbert commented:
Thanks @jwiggins! This was great feedback.

jwiggins commented:
I'd highly recommend adding Travis support in a different PR. You've got plenty of changes going on here already.

MatthieuDartiailh commented:
I still need to polish a bit. I will add some minimal tests before merging and consolidate the test and add Travis in another one.

jwiggins commented on setup.py:
```diff
@@ -62,14 +66,15 @@ def build_extensions(self):
 setup(
     name='atom',
-    version='0.3.10',
+    version='0.3.11',
```
`'0.4.0'`

MatthieuDartiailh commented:
I addressed all comments and ran some basic sanity checks on the examples. I also reverted to using native ints in intenum. I will now work on adding Travis in another PR and merge it first before merging this one. So that we have a proper feedback on it. Compared to the branch I worked for nearly a year the largest change is the handling of integers under Python 3 (so if anybody want to double check, he is welcome).

MatthieuDartiailh commented:
After quite some tinkering (my Linux VM for python 2 broke so I could not test on Python 2) everything seems to be in order I will merge in a week if nobody complains.

MatthieuDartiailh commented:
@sccolbert atom is now ready for a new release ! 0.4.0
You just need to update the releasenotes
